src/main/drivers/sonar_hcsr04.c

```diff
@@ -86,79 +96,136 @@ void hcsr04_init(void)
 void hcsr04_start_reading(void)
 {
 #if !defined(UNIT_TEST)
-     static timeMs_t timeOfLastMeasurementMs = 0;
-    // the firing interval of the trigger signal should be greater than 60ms
-    // to avoid interference between consecutive measurements.
-    #define HCSR04_MinimumFiringIntervalMs 60
+#ifdef SONAR_TRIG_INVERTED
+    IOLo(triggerIO);
+    delayMicroseconds(11);
+    IOHi(triggerIO);
+#else
+    IOHi(triggerIO);
+    delayMicroseconds(11);
+    IOLo(triggerIO);
+#endif
+#endif
+}
+
+void hcsr04_update(void)
+{
     const timeMs_t timeNowMs = millis();
-    if (timeNowMs > timeOfLastMeasurementMs + HCSR04_MinimumFiringIntervalMs) {
-        timeOfLastMeasurementMs = timeNowMs;
-        IOHi(triggerIO);
-        //  The width of trigger signal must be greater than 10us, according to device spec
-        delayMicroseconds(11);
-        IOLo(triggerIO);
+
+    /* the firing interval of the trigger signal should be greater than 60ms
+     * to avoid interference between consecutive measurements */
+    if (timeNowMs > lastMeasurementStartedAt + HCSR04_MinimumFiringIntervalMs) {
+        /* We should have a valid measurement within 60ms of trigger */
+        if ((lastMeasurementReceivedAt - lastMeasurementStartedAt) <= HCSR04_MinimumFiringIntervalMs) {
+            /* 
+             * The speed of sound is 340 m/s or approx. 29 microseconds per centimeter.
+             * The ping travels out and back, so to find the distance of the
+             * object we take half of the distance traveled.
+             *
+             * 340 m/s = 0.034 cm/microsecond = 29.41176471 *2 = 58.82352941 rounded to 59 
+             */
+
+            lastCalculatedDistance = hcsr04SonarPulseTravelTime / 59;
+            if (lastCalculatedDistance > HCSR04_MAX_RANGE_CM) {
+                lastCalculatedDistance = RANGEFINDER_OUT_OF_RANGE;
+            }
+        }
+        else {
+            /* No measurement within reasonable time - indicate failure */
```

Comment:
I prefer '//' for in-code comments, and `/*` for comments at the top of a function.